lib/assert.js

```diff
@@ -287,6 +292,7 @@ assert.notEqual = function notEqual(actual, expected, message) {
 
 // The equivalence assertion tests a deep equality relation.
 assert.deepEqual = function deepEqual(actual, expected, message) {
+  if (isDeepEqual === undefined) lazyLoadComparison();
```

Comment:
Instead of this, you could assign `isDeepEqual` a function that loads and then overwrites itself. That way you can be sure you caught all the use sites.

Reply:
You would have to use `lazyComparison()(args)` the way you defined it. Inlining shouldn't be an issue here.

I'd even go a bit further and define a wrapper that can be reused:

```js
function lazify(loader) {
  let loaded = undefined;
  if (loaded === undefined) loaded = loader();
  return function(...args) { return loaded(...args); };
}
```

You can then use this to load `comparison`:

```js
let comparison = lazify(() => require('internal/util/comparisons'));
```

I wonder though whether inlining that would be an issue. @bmeurer

Reply:
I just ran benchmarks on all three of the variations and looking at the numbers it seems like everything gets inlined (both ways). Depending on how I write the benchmark, the outcome is favorable for some other different versions.

```js
'use strict';

let foo;

let tmpBaz;
function baz() {
  if (tmpBaz === undefined) tmpBaz = require('util');
  return tmpBaz;
}

const lazyUtil = lazify(() => require('util'));
function lazify(loader) {
  let loaded;
  return function() {
    if (loaded === undefined) {
      loaded = loader();
      loader = undefined;
    }
    return loaded;
  };
}

function a(a, b) {
  if (foo === undefined) foo = require('util');
  return foo.isDeepStrictEqual(a, b);
}

function b(a, b) {
  return baz().isDeepStrictEqual(a, b);
}

function c(a, b) {
  return lazyUtil().isDeepStrictEqual(a, b);
}

function bench(fn) {
  const arr = [];
  arr.push(fn(0, 0));
  arr.push(fn(1, 1));
  arr.push(fn(2, 2));
  arr.push(fn(3, 3));
  console.time(`Runtime ${fn.name}`);
  for (var i = 0; i < 1e7; i++) {
    // if (i % 1e4 === 0) {
    //   foo = undefined;
    //   tmpBaz = undefined;
    // }
    arr.push(fn(i, i));
  }
  console.timeEnd(`Runtime ${fn.name}`);
  return arr;
}

for (var i = 0; i < 5; i++) {
  bench(a);
  bench(b);
  bench(c);
}
```

Without the if block, `a` is the winner with ~450ms vs ~500ms. With the if block, `b` is the winner with ~450ms vs ~500ms.

I personally normally use it the way as is right now but I see the point for it being error prone. Even though it also some times allows to only lazy load once outside of a loop instead of running a function in a hot loop.

Reply:
I guess we should agree on one specific way how to do this. I do not have a strong opinion and would like to let others decide which one to pick.

Reply:
I like Yang's way if it is fast.